routes/runtime/github.rb

```diff
@@ -108,36 +108,43 @@ class Clover
+        end
 
-          # Token creation on Cloudflare R2 takes time to propagate. Since that point is the
-          # first time we use the credentials, we are waiting it to be propagated. Note that,
-          # credential propagation will happen only while the bucket and token are being created
-          # initially. So, the retry block expected to run only while saving the first cache
-          # entry for a repository.
-          retries = 0
-          begin
-            upload_id = repository.blob_storage_client.create_multipart_upload(bucket: repository.bucket_name, key: entry.blob_key).upload_id
-          rescue Aws::S3::Errors::Unauthorized, Aws::S3::Errors::InternalError, Aws::S3::Errors::NoSuchBucket => ex
-            retries += 1
-            if retries < 3
-              # :nocov:
-              sleep(1) unless Config.test?
-              # :nocov:
-              retry
-            else
-              Clog.emit("Could not authorize multipart upload") { {could_not_authorize_multipart_upload: {ubid: runner.ubid, repository_ubid: repository.ubid, exception: Util.exception_to_hash(ex)}} }
-              fail CloverError.new(400, "InvalidRequest", "Could not authorize multipart upload")
-            end
+        # Need id for blob_key, but don't save record yet
+        entry = GithubCacheEntry.new_with_id(repository_id: runner.repository.id, key:, version:, size:, scope:, created_by: runner.id)
+        blob_key = entry.blob_key
+        bucket = repository.bucket_name
+        blob_storage_client = repository.blob_storage_client
+
+        # Token creation on Cloudflare R2 takes time to propagate. Since that point is the
+        # first time we use the credentials, we are waiting it to be propagated. Note that,
+        # credential propagation will happen only while the bucket and token are being created
+        # initially. So, the retry block expected to run only while saving the first cache
+        # entry for a repository.
+        retries = 0
+        begin
+          upload_id = blob_storage_client.create_multipart_upload(bucket:, key: blob_key).upload_id
+        rescue Aws::S3::Errors::Unauthorized, Aws::S3::Errors::InternalError, Aws::S3::Errors::NoSuchBucket => ex
+          retries += 1
+          if retries < 3
+            # :nocov:
+            sleep(1) unless Config.test?
+            # :nocov:
+            retry
+          else
+            Clog.emit("Could not authorize multipart upload") { {could_not_authorize_multipart_upload: {ubid: runner.ubid, repository_ubid: repository.ubid, exception: Util.exception_to_hash(ex)}} }
+            fail CloverError.new(400, "InvalidRequest", "Could not authorize multipart upload")
           end
+        end
 
+        begin
           entry.update(upload_id:)
+        rescue Sequel::ValidationFailed, Sequel::UniqueConstraintViolation
+          blob_storage_client.delete_object(bucket:, key: blob_key)
```

Comment:
Since we haven't completed the multipart upload, the object hasn't been created yet. So, `delete_object` might not work as expected. What we want is `abort_multipart_upload`.

Also, Cloudflare R2 aborts any incomplete multipart uploads after 24 hours.

Reply:
I removed this line and rebased this PR on main.  I also removed the second commit, since it isn't helpful if we aren't calling delete_object in multiple spots.